libraries/chain/include/graphene/chain/event_object.hpp

```diff
@@ -0,0 +1,155 @@
+#include <graphene/db/object.hpp>
+#include <graphene/db/generic_index.hpp>
+#include <graphene/chain/protocol/event.hpp>
+
+namespace graphene { namespace chain {
+   class event_object;
+} }
+
+namespace fc { 
+   void to_variant(const graphene::chain::event_object& event_obj, fc::variant& v);
+   void from_variant(const fc::variant& v, graphene::chain::event_object& event_obj);
+} //end namespace fc
+
+namespace graphene { namespace chain {
+
+class database;
+
+class event_object : public graphene::db::abstract_object< event_object >
+{
+   public:
+      static const uint8_t space_id = protocol_ids;
+      static const uint8_t type_id = event_object_type;
+
+      event_object();
+      event_object(const event_object& rhs);
+      ~event_object();
+      event_object& operator=(const event_object& rhs);
+
+      internationalized_string_type name;
+      
+      internationalized_string_type season;
+
+      optional<time_point_sec> start_time;
+
+      event_group_id_type event_group_id;
+
+      bool at_least_one_betting_market_group_settled;
+
+      event_status get_status() const;
+      vector<string> scores;
+
+      // serialization functions:
+      // for serializing to raw, go through a temporary sstream object to avoid
+      // having to implement serialization in the header file
+      template<typename Stream>
+      friend Stream& operator<<( Stream& s, const event_object& event_obj );
+
+      template<typename Stream>
+      friend Stream& operator>>( Stream& s, event_object& event_obj );
+
+      friend void ::fc::to_variant(const graphene::chain::event_object& event_obj, fc::variant& v);
+      friend void ::fc::from_variant(const fc::variant& v, graphene::chain::event_object& event_obj);
+
+      void pack_impl(std::ostream& stream) const;
+      void unpack_impl(std::istream& stream);
+
+      void on_upcoming_event(database& db);
+      void on_in_progress_event(database& db);
+      void on_frozen_event(database& db);
+      void on_finished_event(database& db);
+      void on_canceled_event(database& db);
+      void on_settled_event(database& db);
+      void on_betting_market_group_resolved(database& db, betting_market_group_id_type resolved_group, bool was_canceled);
+      void on_betting_market_group_closed(database& db, betting_market_group_id_type closed_group);
+      void dispatch_new_status(database& db, event_status new_status);
+   private:
+      class impl;
+      std::unique_ptr<impl> my;
+};
+
+struct by_event_group_id;
+struct by_event_status;
+typedef multi_index_container<
+   event_object,
+   indexed_by<
+      ordered_unique< tag<by_id>, member< object, object_id_type, &object::id > >,
+      ordered_non_unique< tag<by_event_group_id>, member< event_object, event_group_id_type, &event_object::event_group_id > >,
+      ordered_non_unique< tag<by_event_status>, const_mem_fun< event_object, event_status, &event_object::get_status > > > > event_object_multi_index_type;
```

Comment:
This should be made `ordered_unique` to enforce a well-defined order when iterating.